lib/msal-node/src/network/HttpClient.ts

```diff
@@ -60,29 +43,255 @@ export class HttpClient implements INetworkModule {
+                });
+                const headers = Object.fromEntries(entries);
+
+                const networkResponse: NetworkResponse<T> = {
+                    headers: headers as Record<string, string>,
+                    body: JSON.parse(body) as T,
+                    status: statusCode as number,
+                };
+
+                if ((statusCode < 200 || statusCode > 299) &&
+                    // do not destroy the request for the device code flow
+                    networkResponse.body["error"] !== Constants.AUTHORIZATION_PENDING) {
+                    request.destroy();
+                    socket.destroy();
+                    reject(new Error(`HTTP status code ${statusCode}`));
+                }
+
+                resolve(networkResponse);
+            });
+
+            socket.on("error", (chunk) => {
+                request.destroy();
+                socket.destroy();
+                reject(new Error(chunk.toString()));
+            });
+        });
+
+        request.on("error", (chunk) => {
+            request.destroy();
+            reject(new Error(chunk.toString()));
+        });
+    }));
+};
+
+const networkRequestViaHttps = <T>(
+    url: string,
+    httpMethod: string,
+    options?: NetworkRequestOptions,
+    timeout?: number,
+): Promise<NetworkResponse<T>> => {
+    const isPostRequest = httpMethod === HttpMethod.POST;
+    const body: string = options?.body || "";
+
+    const emptyHeaders: Record<string, string> = {};
+    const customOptions: https.RequestOptions = {
+        method: httpMethod,
+        headers: options?.headers || emptyHeaders,
+    };
+
+    if (timeout) {
+        customOptions.timeout = timeout;
+    }
+
+    if (isPostRequest) {
+        // needed for post request to work
+        customOptions.headers = {
+            ...customOptions.headers,
+            "Content-Length": body.length,
         };
     }
-}
+
+    return new Promise<NetworkResponse<T>>((resolve, reject) => {
+        const request = https.request(url, customOptions);
+
+        if (timeout) {
+            request.on("timeout", () => {
+                request.destroy();
+                reject(new Error("Request time out"));
+            });
+        }
+
+        if (isPostRequest) {
+            request.write(body);
+        }
+
+        request.end();
+
+        request.on("response", (response) => {
+            const headers = response.headers;
+            const statusCode = response.statusCode as number;
+
+            const data: Buffer[] = [];
+            response.on("data", (chunk) => {
+                data.push(chunk);
+            });
+
+            response.on("end", () => {
+                // combine all received buffer streams into one buffer, and then into a string
+                let body = Buffer.concat([...data]).toString();
+
+                // check if the value of the body is already a JSON object
+                try {
+                    JSON.parse(body);
```

Comment:
Instead of potentially calling `JSON.parse` twice (which is an expensive operation), can you store this value if it is successful, and use it below?

Reply:
I think I had put this additional logic in thinking it would fix device flow, before I discovered what the actual problem was. I've reverted this code.